File: problems/sorting/LC135_candy/solution.py
class Solution:
  def candy(self, ratings):
    n = len(ratings)
    pref = [1]*n
    suff = [1]*n
    for i in range(1, n):
        if ratings[i] > ratings[i - 1]:
            pref[i] = 1 + pref[i - 1]
    for i in range(n - 2, -1, -1):
        if ratings[i] > ratings[i + 1]:
            suff[i] = 1 + suff[i + 1]
    ans = 0
    for i in range(n):
        ans += max(pref[i], suff[i])
    return ans
  
def main():
  ratings = [1,0,2]
  s = Solution()
  print(s.candy(ratings))

main()

File: problems/sorting/LC135_candy/solution.go
package main

import (
	"fmt"
)

func candy(ratings []int) int {
	n := len(ratings)
	pref := make([]int, n)
	suff := make([]int, n)
	for i := range pref {
		pref[i] = 1
		suff[i] = 1
	}
	for i := 1; i < n; i++ {
		if ratings[i] > ratings[i-1] {
			pref[i] = 1 + pref[i-1]
		}
	}
	for i := n - 2; i >= 0; i-- {
		if ratings[i] > ratings[i+1] {
			suff[i] = 1 + suff[i+1]
		}
	}
	ans := 0
	for i := 0; i < n; i++ {
		ans += max(pref[i], suff[i])
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	ratings := []int{1, 0, 2}
	fmt.Println(candy(ratings))
}
